### PR TITLE
Universal streamk with atomics

### PR DESCRIPTION
**Pull request by hsadasiv**
universal_streamk with atomics is implemented & integrated with ckprofiler. Lot of the modified files are ckprofiler instances and need only be skimmed through. 
universal_streamk supports 1-tile and 2-tile stream-k and programmable grid_size (that defaults to max occupancy) for a persistent kernel. 
As PR is already huge, a second PR will be submitted later for supporting stream-k with reduction and advanced tile sizzling mechanisms.
Example is added to readme for better understanding of usage: https://github.com/ROCm/composable_kernel/blob/universal_streamk/example/01_gemm/README.md

### CODE REVIEW

**Comment by aosewski** [comment]
@hsadasiv  I believe you've unintentionally changed file permissions: 
obraz

File: example/01_gemm/gemm_xdl_fp16_streamk_v3.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (aosewski) [inline example/01_gemm/gemm_xdl_fp16_streamk_v3.cpp]: You have to update copyright year. It's handy to use `precommit` (see repo readme for instructions) to not overlook such things.

File: example/01_gemm/run_gemm_example_streamk_v2.inc [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    else if constexpr(std::is_same_v<DataType, ck::bf8_t>)
+    {
+        return 1.5e-1; // 57344 and 49152 are acceptable
+    }
+    else
+    {
+        return 1e-3;
+    }
+}
```
Comment (aosewski) [inline example/01_gemm/run_gemm_example_streamk_v2.inc:45]: This looks really nice. Maybe you could integrate it with  `check_err` function ?

File: example/01_gemm/run_gemm_example_streamk_v2.inc [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+template <typename DataType>
+inline __host__ __device__ constexpr double get_atol()
```
Comment (aosewski) [inline example/01_gemm/run_gemm_example_streamk_v2.inc:48]: As above.

File: example/01_gemm/run_gemm_example_streamk_v2.inc [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            }
+            else
+            {
+                return HostTensorDescriptor({row, col}, {1_uz, stride});
+            }
+        };
+
+    auto f_get_default_stride =
+        [](std::size_t row, std::size_t col, std::size_t stride, auto layout) {
+            if(stride == 0)
```
Comment (aosewski) [inline example/01_gemm/run_gemm_example_streamk_v2.inc]: Just consider: (examples/ckProfiler usage consistency) I think we have "unwritten" rule to use `-1` for default stride values.

File: example/01_gemm/run_gemm_example_streamk_v2.inc [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    for (int in = 0; in < N; in++)
+    {
+        for (int ik = 0; ik < K; ik++)
+        {
+            printf("%02x ", *(reinterpret_cast<uint8_t*>(&b_k_n(ik,in))));
+            if(ik%8==7) printf("|");
+        }
+        printf("\n");
+    }
+#endif
```
Comment (aosewski) [inline example/01_gemm/run_gemm_example_streamk_v2.inc]: This looks like debugging leftover.

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+            float ave_time = 0;
+
+            index_t k_grain = KPerBlock;
+            index_t K_split = (arg.K + k_grain - 1) / k_grain * KPerBlock;
+
+            const bool has_main_k_block_loop = GridwiseGemm::CalculateHasMainKBlockLoop(K_split);
+            hipGetErrorString(hipMemsetAsync(arg.p_c_grid,
+                                             0,
+                                             arg.M * arg.N * sizeof(CDataType),
+                                             stream_config.stream_id_)); // HS
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: What is this comment about?

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+                        // rotating mem
+                        rotating_mem.Next();
+                    };
+
+                    ave_time = ck::utility::launch_and_time_kernel_with_preprocess<false>(
+                        stream_config,
+                        run_flush_cache,
+                        kernel,
+                        grid_dim,
+                        // dim3(gdx, gdy, gdz),
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: leftover?

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+                        dim3(BlockSize),
+                        0,
+                        arg_);
+                }
+                else
+                {
+
+                    ave_time = launch_and_time_kernel(stream_config,
+                                                      kernel,
+                                                      //   dim3(gdx, gdy, gdz),
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: leftover

File: library/src/tensor_operation_instance/gpu/gemm_universal_streamk/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_comp_mnkpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v1_default_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v1_kpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v1_mnkpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v2_default_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v2_kpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v2_mnkpadding_instance.cpp
+        
+        # device_gemm_xdl_universal_f16_f8_f16/device_gemm_xdl_universal_f16_f8_f16_mk_kn_mn_comp_default_instance.cpp
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/gemm_universal_streamk/CMakeLists.txt]: I guess all commented lines below are unintentional.

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+                             index_t StrideC,
+                             index_t streamk_sel,
+                             index_t Grid_size,
+                             AElementwiseOperation,
+                             BElementwiseOperation,
+                             CElementwiseOperation)
+    {
+        // return Argument{p_a, p_b, p_c, M, N, K, StrideA, StrideB, StrideC, KBatch};
+        return Argument{
+            p_a, p_b, p_c, M, N, K, StrideA, StrideB, StrideC, streamk_sel, Grid_size}; // HS
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: Leftover comments.

File: library/include/ck/library/tensor_operation_instance/gpu/gemm_universal_streamk.hpp [hunk 11]
```diff
@@ -0,0 +1,788 @@
...
+                                                      Col,
+                                                      Row,
+                                                      F16,
+                                                      F16,
+                                                      F16,
+                                                      PassThrough,
+                                                      PassThrough,
+                                                      PassThrough>>>& instances);
+#endif
+// #if(defined(CK_ENABLE_FP16) || defined(CK_ENABLE_FP8))
```
Comment (aosewski) [inline library/include/ck/library/tensor_operation_instance/gpu/gemm_universal_streamk.hpp]: Please clean all those commented unused code.

File: library/src/tensor_operation_instance/gpu/gemm_universal_streamk/gemm_universal_streamk/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_comp_mnkpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v1_default_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v1_kpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v1_mnkpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v2_default_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v2_kpadding_instance.cpp
+        device_gemm_xdl_universal_streamk_f16_f16_f16/device_gemm_xdl_universal_streamk_f16_f16_f16_mk_nk_mn_mem_v2_mnkpadding_instance.cpp
+        
+        # device_gemm_xdl_universal_f16_f8_f16/device_gemm_xdl_universal_f16_f8_f16_mk_kn_mn_comp_default_instance.cpp
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/gemm_universal_streamk/gemm_universal_streamk/CMakeLists.txt]: Leftovers

File: profiler/src/profile_gemm_universal_streamk.cpp [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+    if(data_type == GemmDataType::F16_F16_F16 && layout == GemmMatrixLayout::MK_KN_MN)
+    {
+        return profile(F16{}, F16{}, F32{}, F16{}, Row{}, Row{}, Row{});
+    }
+    else if(data_type == GemmDataType::F16_F16_F16 && layout == GemmMatrixLayout::MK_NK_MN)
+    {
+        return profile(F16{}, F16{}, F32{}, F16{}, Row{}, Col{}, Row{});
+    }
+    // else if(data_type == GemmDataType::F16_F8_F16 && layout == GemmMatrixLayout::MK_KN_MN)
```
Comment (aosewski) [inline profiler/src/profile_gemm_universal_streamk.cpp]: Leftovers.

File: profiler/include/profiler/profile_gemm_universal_streamk_impl.hpp [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+    float best_tflops      = 0;
+    float best_gb_per_sec  = 0;
+    float best_grid_size   = 0;
+    float best_streamk_sel = 0;
+
+    // profile device GEMM instances
+    for(auto& op_ptr : op_ptrs)
+    {
+        std::vector<int> grid_size_list   = {38, 76, 114, 152, 190, 228, 266, 304, 342, 380};
+        std::vector<int> streamk_sel_list = {0, 1, 2, 3, 4};
```
Comment (aosewski) [inline profiler/include/profiler/profile_gemm_universal_streamk_impl.hpp]: What would mean values greater than 2 ?

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 15]
```diff
@@ -0,0 +1,2063 @@
...
+    // Pass two lds pointer is the key to tell compiler that ds_read/write
+    // operate on different lds chunk at same time without order dependecy
+    __shared__ char p_shared_0[GridwiseGemm::GetSharedMemoryNumberOfByte()];
+    __shared__ char p_shared_1[GridwiseGemm::GetSharedMemoryNumberOfByte()];
+
+    // auto splitk_batch_offset = typename GridwiseGemm::SplitKBatchOffset(karg);
+
+    GridwiseGemm::template Run_2Lds<HasMainKBlockLoop, CGlobalMemoryDataOperation, TailNum>(
+        // karg.p_a_grid + splitk_batch_offset.a_k_split_offset,
+        // karg.p_b_grid + splitk_batch_offset.b_k_split_offset,
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_streamk_v3.hpp]: Leftovers.

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 16]
```diff
@@ -0,0 +1,2063 @@
...
+    kernel_gemm_xdl_cshuffle_v3(typename GridwiseGemm::Argument karg)
+{
+#if(!defined(__HIP_DEVICE_COMPILE__) || defined(__gfx9__))
+    __shared__ char p_shared[GridwiseGemm::GetSharedMemoryNumberOfByte()];
+
+    // auto splitk_batch_offset = typename GridwiseGemm::SplitKBatchOffset(karg);
+
+    GridwiseGemm::template Run<HasMainKBlockLoop, CGlobalMemoryDataOperation, TailNum>(
+        // karg.p_a_grid + splitk_batch_offset.a_k_split_offset,
+        // karg.p_b_grid + splitk_batch_offset.b_k_split_offset,
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_streamk_v3.hpp]: Leftovers.

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_streamk_v3.hpp [hunk 17]
```diff
@@ -0,0 +1,2063 @@
...
+    static constexpr auto AK1Number = Number<AK1Value>{};
+    static constexpr auto BK1Number = Number<BK1Value>{};
+
+    static constexpr index_t KPack =
+        math::max(math::lcm(AK1Number, BK1Number),
+                  MfmaSelector<ComputeTypeA, MPerXdl, NPerXdl>::selected_mfma.k_per_blk);
+
+    using ThisThreadBlock = ThisThreadBlock<BlockSize>;
+
+    // __host__ static auto CalculateGridSize(index_t M, index_t N) //, index_t KBatch)
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_streamk_v3.hpp]: Leftover

File: include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp [hunk 18]
```diff
@@ -1404,4 +1404,440 @@ struct BlockToCTileMap_GemmStreamK
...
+
+        if constexpr(ReductionStrategy == StreamKReductionStrategy::Reduction)
+        {
+            uint32_t upper_big    = math::lcm(k_iters_per_big_block, k_iters_per_tile.get());
+            uint32_t upper_little = math::lcm(k_iters_per_big_block - 1, k_iters_per_tile.get());
+            equiv_tiles_big       = MDiv(upper_big / k_iters_per_tile.get());
+            equiv_tiles_little    = MDiv(upper_little / k_iters_per_tile.get());
+        }
+
+#if 0
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp]: Maybe we could use ` if(ck::EnvIsEnabled(CK_ENV(CK_LOGGING)))` here ?

**Review by aosewski — COMMENTED** [review]
Haven't yet reviewed gridwise gemm Run function for 2LDS and haven't yet finished block-to-c-tile-map. In general looks good, but there's a loooot of leftover commented code which need to be cleaned-up. 

1. Please update copyright year in all headers to 2024 if needed. 
2. Remove all commented out code.

File: example/01_gemm/run_gemm_example_streamk_v2.inc (hunk 2 above)
Comment (hsadasiv) [reply example/01_gemm/run_gemm_example_streamk_v2.inc:45]: Good thought! But this is not new and already exists in https://github.com/ROCm/composable_kernel/blob/develop/example/01_gemm/run_gemm_example_v2.inc.
I think merging to check_err requires a clean up of the develop branch for atleast 3 different kernels. So, I'd leave it for a different PR

File: example/01_gemm/run_gemm_example_streamk_v2.inc (hunk 3 above)
Comment (hsadasiv) [reply example/01_gemm/run_gemm_example_streamk_v2.inc:48]: This is not a new piece of code introduced in universal_streamk and is used in 3 different kernels in develop branch including streamk. This requires clean-up of other files and is best left for a different PR I believe.

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp (hunk 10 above)
Comment (hsadasiv) [reply include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: removed

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp (hunk 6 above)
Comment (hsadasiv) [reply include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: removed.

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp (hunk 7 above)
Comment (hsadasiv) [reply include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: removed

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp (hunk 8 above)
Comment (hsadasiv) [reply include/ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_streamk_v3.hpp]: This seems to be a valid comment forwarded from streamk (merged to develop) to universal stream-k.

File: example/01_gemm/run_gemm_example_streamk_v2.inc (hunk 4 above)
Comment (hsadasiv) [reply example/01_gemm/run_gemm_example_streamk_v2.inc]: Great point! I will correct it.

File: example/01_gemm/run_gemm_example_streamk_v2.inc (hunk 5 above)
Comment (hsadasiv) [reply example/01_gemm/run_gemm_example_streamk_v2.inc]: Seems it got forwarded from old stream-k. Will remove it.

File: library/src/tensor_operation_instance/gpu/gemm_universal_streamk/gemm_universal_streamk/CMakeLists.txt (hunk 12 above)
Comment (hsadasiv) [reply library/src/tensor_operation_instance/gpu/gemm_universal_streamk/gemm_universal_streamk/CMakeLists.txt]: Removed. Thanks.

File: library/src/tensor_operation_instance/gpu/gemm_universal_streamk/CMakeLists.txt (hunk 9 above)
Comment (hsadasiv) [reply library/src/tensor_operation_instance/gpu/gemm_universal_streamk/CMakeLists.txt]: Removed. thanks.

File: profiler/include/profiler/profile_gemm_universal_streamk_impl.hpp (hunk 14 above)
Comment (hsadasiv) [reply profiler/include/profiler/profile_gemm_universal_streamk_impl.hpp]: comments added. thanks.

File: profiler/src/profile_gemm_universal_streamk.cpp (hunk 13 above)
Comment (hsadasiv) [reply profiler/src/profile_gemm_universal_streamk.cpp]: removed. thanks.

File: library/include/ck/library/tensor_operation_instance/gpu/gemm_universal_streamk.hpp (hunk 11 above)
Comment (hsadasiv) [reply library/include/ck/library/tensor_operation_instance/gpu/gemm_universal_streamk.hpp]: Done. Thanks.

File: example/01_gemm/gemm_xdl_fp16_streamk_v3.cpp (hunk 1 above)
Comment (hsadasiv) [reply example/01_gemm/gemm_xdl_fp16_streamk_v3.cpp]: fixed. Thanks for the suggestion. Passes precommit checks now.

**Comment by hsadasiv** [comment]
@aosewski 
Thank You for your detailed review. I appreciate your effort!
I have tried to address all of them. Please check and approve.

File: .pre-commit-config.yaml
Comment (aosewski) [inline .pre-commit-config.yaml:1]: Incorrect file permissions sneaked-into this file

**Review by aosewski — APPROVED** [review]

**Pull request merged**